platform/cog-platform-fdo.c

```diff
@@ -1266,6 +1392,130 @@ on_export_fdo_egl_image(void *data, struct wpe_fdo_egl_exported_image *image)
     wl_surface_commit (win_data.wl_surface);
 }
 
+#if COG_ENABLE_WESTON_DIRECT_DISPLAY
+static void
+create_succeeded(void *data,
+		 struct zwp_linux_buffer_params_v1 *params,
+		 struct wl_buffer *new_buffer)
+{
+	zwp_linux_buffer_params_v1_destroy (params);
+
+	struct buffer *buffer = data;
+	buffer->buffer = new_buffer;
+}
+
+static void
+create_failed(void *data, struct zwp_linux_buffer_params_v1 *params)
+{
+    struct buffer *buffer = data;
+
+    buffer->buffer = NULL;
+    zwp_linux_buffer_params_v1_destroy (params);
+    g_assert_not_reached ();
+}
+
+static const struct zwp_linux_buffer_params_v1_listener params_listener = {
+	create_succeeded,
+	create_failed
+};
+
+static void
+destroy_video_surface (gpointer data)
+{
+    video_surface *surface = (video_surface*) data;
+
+    g_clear_pointer (&surface->protected_surface, weston_protected_surface_destroy);
+    g_clear_pointer (&surface->wl_subsurface, wl_subsurface_destroy);
+    g_clear_pointer (&surface->wl_surface, wl_surface_destroy);
+
+    free (surface);
+}
+
+static void
+on_video_plane_display_dmabuf_receiver_handle_dmabuf (void* data, struct wpe_video_plane_display_dmabuf_export* dmabuf_export, uint32_t id, int fd, int32_t x, int32_t y, int32_t width, int32_t height, uint32_t stride)
+{
+    if (fd < 0)
+        return;
+
+    if (!wl_data.dmabuf) {
+        // TODO: Replace with g_warning_once() after bumping our GLib requirement.
+        static bool warning_emitted = false;
+        if (!warning_emitted) {
+            g_warning ("DMABuf not supported by the compositor. Video won't be rendered");
+            warning_emitted = true;
+        }
+        return;
+    }
+
+    uint64_t modifier = DRM_FORMAT_MOD_INVALID;
+    struct zwp_linux_buffer_params_v1 *params = zwp_linux_dmabuf_v1_create_params (wl_data.dmabuf);
+    if (wl_data.direct_display != NULL)
+        weston_direct_display_v1_enable (wl_data.direct_display, params);
+
+    video_surface *surf = (video_surface*) g_hash_table_lookup (win_data.video_surfaces, GUINT_TO_POINTER(id));
+    if (!surf) {
+        surf = calloc (1, sizeof (video_surface));
```

Comment:
How about using `g_slice_new0()` here? 

```c
surf = g_slice_new0 (video_surface);
```

And of course `g_slice_free()` where the `video_surface` is deallocated.